### PR TITLE
sql: rearchitecture ALTER TABLE RENAME, add support for renaming constraints

### LINKED ISSUE #32555: sql: support ALTER TABLE ... RENAME CONSTRAINT
**Describe the solution you'd like**

Enable renaming constraints with `ALTER TABLE ... RENAME CONSTRAINT`

https://www.postgresql.org/docs/10/sql-altertable.html

### PR DESCRIPTION
**Pull request by knz**
Fixes #32555. For TypeORM compat, see discussion on #22298.

Release note (sql change): This patch changes RENAME COLUMN to become
a "table command", which can be used alongside other table commands in
a single ALTER TABLE statement. This makes it possible to e.g. atomically
add a computed column based on an existing column, and rename the
columns so that the computed column "replaces" the original column.

Release note (sql change): CockroachDB now supports ALTER TABLE RENAME
CONSTRAINT for compatibility with PostgreSQL. This feature is limited
to *named* constraints, where the name of the constraints is preserved
in the table metadata. This currently includes CHECK, UNIQUE and
FOREIGN KEY constraints, and does not include other
constraints (DEFAULT, NULL etc) otherwise supported by PostgreSQL. For
UNIQUE constraint, only supporting indexes that are not depended on by
views can be renamed.

### CODE REVIEW

**Comment by knz** [comment]
ping

**Comment by dt** [comment]
cc @lucy-zhang who's actively working on some constraint-related schema mgmt stuff right now

**Review by thoszhang — COMMENTED** [review]
I'll review this more thoroughly later, but I have one question for now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35091#-:-LZaHXxw5eYNonEaCqsL:bxw2ess)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bobvawter, @BramGruneir, @dt, @knz, and @lucy-zhang)

___
*[pkg/sql/sqlbase/structured.go, line 2110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35091#-LZaHTwHCdj3Ksz5-ngz:-LZaHTwHCdj3Ksz5-nh-:b6cddce) ([raw file](https://github.com/cockroachdb/cockroach/blob/d005deef4887150d9faa1c8b8ebef54b12409fb0/pkg/sql/sqlbase/structured.go#L2110)):*
> ```Go
> 
> // RenameConstraint renames a constraint.
> func (desc *MutableTableDescriptor) RenameConstraint(
> ```

I haven't tested this myself yet, but what happens if a constraint is renamed while it's being added? We now enqueue a mutation to validate check constraints when they're added, and the mutation only stores the name of the check constraint (since they don't have IDs, which is unfortunate), so when the validation runs, I think it will fail if the constraint has been renamed in the meantime.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35091)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bobvawter, @BramGruneir, @dt, and @lucy-zhang)

___
*[pkg/sql/sqlbase/structured.go, line 2110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35091#-LZaHTwHCdj3Ksz5-ngz:-LZaTh7l7xR05IESHV-o:b-oayvs3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d005deef4887150d9faa1c8b8ebef54b12409fb0/pkg/sql/sqlbase/structured.go#L2110)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I haven't tested this myself yet, but what happens if a constraint is renamed while it's being added? We now enqueue a mutation to validate check constraints when they're added, and the mutation only stores the name of the check constraint (since they don't have IDs, which is unfortunate), so when the validation runs, I think it will fail if the constraint has been renamed in the meantime.
</blockquote></details>

I don't have an answer for this. You're right it may break. I can add a TODO, file an issue, and add an extra error check in the constraint validation code in case "the name does not exist any more". I think the case is sufficiently rare that an error in that case is ok for now. What do you think?

(Alternatively if you have an idea of a solution, I'll take it)

**Review by thoszhang — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35091)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bobvawter, @BramGruneir, @dt, and @knz)

___
*[pkg/sql/sqlbase/structured.go, line 2110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35091#-LZaHTwHCdj3Ksz5-ngz:-LZam5Un4ASSlfGZJiZg:bhfnki5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d005deef4887150d9faa1c8b8ebef54b12409fb0/pkg/sql/sqlbase/structured.go#L2110)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't have an answer for this. You're right it may break. I can add a TODO, file an issue, and add an extra error check in the constraint validation code in case "the name does not exist any more". I think the case is sufficiently rare that an error in that case is ok for now. What do you think?

(Alternatively if you have an idea of a solution, I'll take it)
</blockquote></details>

Could we disallow renaming constraints that are being added? We currently disallow deleting constraints being added: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/alter_table.go#L466-L468

**Review by thoszhang — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35091)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bobvawter, @BramGruneir, @dt, and @knz)

___
*[pkg/sql/sqlbase/structured.go, line 2110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35091#-LZaHTwHCdj3Ksz5-ngz:-LZao5noEyb1CpjwGgBD:b30nt2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d005deef4887150d9faa1c8b8ebef54b12409fb0/pkg/sql/sqlbase/structured.go#L2110)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Could we disallow renaming constraints that are being added? We currently disallow deleting constraints being added: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/alter_table.go#L466-L468
</blockquote></details>

This is just for check constraints. I don't think FKs (and other constraints) have this problem, since I haven't made changes to FKs yet.

**Review by bobvawter — APPROVED** [review]
I'll second Lucy's point about dealing with renames in the face of a long-running schema change.  I think it's certainly reasonable to kick back an error message to the user saying that the proposed change conflicts with a pending schema mutation and to retry later.

> Reviewed 14 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35091#-:-LZe26Jc0q7H8dgFegp0:b-nbg7v2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @BramGruneir, @dt, and @knz)

___
*[pkg/sql/alter_table.go, line 609 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35091#-LZe1CEH-JKvbrxOwAZH:-LZe1CEI0VhiKz5mbhut:bmncmfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d005deef4887150d9faa1c8b8ebef54b12409fb0/pkg/sql/alter_table.go#L609)):*
> ```Go
> 
> 		case *tree.AlterTableRenameColumn:
> 			descChanged, err := params.p.renameColumn(params.ctx, n.tableDesc, &t.Column, &t.NewName)
> ```

Does this need a privilege check like the constraint rename below, or is that handled elsewhere in the more general `alter table` flow?

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35091)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bobvawter, @BramGruneir, @dt, and @lucy-zhang)

___
*[pkg/sql/alter_table.go, line 609 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35091#-LZe1CEH-JKvbrxOwAZH:-LZfVEsz4jMG7oJS4JKm:b-7iadwx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d005deef4887150d9faa1c8b8ebef54b12409fb0/pkg/sql/alter_table.go#L609)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Does this need a privilege check like the constraint rename below, or is that handled elsewhere in the more general `alter table` flow?
</blockquote></details>

The privilege check is made inside `renameColumn()`.
___
*[pkg/sql/sqlbase/structured.go, line 2110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35091#-LZaHTwHCdj3Ksz5-ngz:-LZfZybh7GWDkGKh4giF:b-poxiwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d005deef4887150d9faa1c8b8ebef54b12409fb0/pkg/sql/sqlbase/structured.go#L2110)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

This is just for check constraints. I don't think FKs (and other constraints) have this problem, since I haven't made changes to FKs yet.
</blockquote></details>

Ok thanks I have added the check. Can you hand-hold me as to where I can put a test for this? Thanks

**Review by thoszhang — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35091)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bobvawter, @BramGruneir, @dt, and @knz)

___
*[pkg/sql/sqlbase/structured.go, line 2110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35091#-LZaHTwHCdj3Ksz5-ngz:-LZk535H0NfXoH6mFkQi:btsz701) ([raw file](https://github.com/cockroachdb/cockroach/blob/d005deef4887150d9faa1c8b8ebef54b12409fb0/pkg/sql/sqlbase/structured.go#L2110)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Ok thanks I have added the check. Can you hand-hold me as to where I can put a test for this? Thanks
</blockquote></details>

We have some logic tests for the case where the disallowed schema change happens in the same transaction as creating the constraint: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/schema_change_in_txn#L844-L895

There is also a similar test in `schema_changer_test.go` where we test that a column can't be dropped if there is a check constraint being added that references it: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/schema_changer_test.go#L1244

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35091)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bobvawter, @BramGruneir, @dt, and @lucy-zhang)

___
*[pkg/sql/sqlbase/structured.go, line 2110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35091#-LZaHTwHCdj3Ksz5-ngz:-L_ObDxBDET8ms9e7MoG:b5kixqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d005deef4887150d9faa1c8b8ebef54b12409fb0/pkg/sql/sqlbase/structured.go#L2110)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

We have some logic tests for the case where the disallowed schema change happens in the same transaction as creating the constraint: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/schema_change_in_txn#L844-L895

There is also a similar test in `schema_changer_test.go` where we test that a column can't be dropped if there is a check constraint being added that references it: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/schema_changer_test.go#L1244
</blockquote></details>

Added the test. Thanks for your guidance!

File: pkg/sql/logictest/testdata/logic_test/rename_column [hunk 1]
```diff
@@ -128,3 +131,14 @@ SELECT column_name FROM [SHOW COLUMNS FROM users]
...
 species
+
+# Check that a column can be added and renamed in the same statement
+statement ok
+ALTER TABLE users RENAME COLUMN species TO species_old,
+                  ADD COLUMN species STRING AS (species_old || ' woo') STORED
+
+query T
+SELECT species FROM users
+----
```
Comment (dt) [inline pkg/sql/logictest/testdata/logic_test/rename_column]: rowsort

**Pull request merged**